### PR TITLE
Reparse block for historydb if corrupted

### LINKED ISSUE #1638: If historydb has corruption, rebuild it instead of doing a full resync

### PR DESCRIPTION
**Pull request by iketheadore**
Fixes #1638 

Changes:
- Rebuild the blocks to build the corrupted history

Does this change need to mentioned in CHANGELOG.md?
No

### CODE REVIEW

File: src/visor/db.go [hunk 1]
```diff
@@ -29,7 +31,35 @@ type ErrCorruptDB struct {
...
+	cb.lock.Lock()
+	var seqs []uint64
+	for seq := range cb.v {
+		seqs = append(seqs, seq)
+	}
+	cb.lock.Unlock()
+	return seqs
+}
+
+// CheckDatabase checks the database for corruption, rebuild history if corrupted
```
Comment (gz-c) [inline src/visor/db.go]: Caller should choose to do history rebuild based upon corrupted state

Comment (iketheadore) [reply src/visor/db.go]: What is the corrupted state mean? We only rebuild when the error is `ErrHistoryDBCorrupted` type

File: src/visor/db.go [hunk 2]
```diff
@@ -49,45 +51,155 @@ func CheckDatabase(db *dbutil.DB, pubkey cipher.PubKey, quit chan struct{}) erro
...
+		lock.Unlock()
+		return err
 	default:
 		return err
 	}
 }
 
-// ResetCorruptDB checks the database for corruption and if corrupted, then it erases the db and starts over.
+// backup the corrypted db first, then rebuild the history DB.
+func rebuildHistoryDB(db *dbutil.DB, history *historydb.HistoryDB, bc *Blockchain, quit chan struct{}) (*dbutil.DB, error) {
```
Comment (gz-c) [inline src/visor/db.go]: backing up and rebuilding should be separate methods

File: src/visor/db.go [hunk 3]
```diff
@@ -49,45 +51,155 @@ func CheckDatabase(db *dbutil.DB, pubkey cipher.PubKey, quit chan struct{}) erro
...
-		// Verify historydb
-		return history.Verify(tx, b, indexesMap)
+		// Verify historydb, we don't return the error of history.Verify here,
+		// as we have to check all signature, if we return error early here, the
+		// potential bad signature won't be detected.
+		lock.Lock()
+		if historyVerifyErr == nil {
+			historyVerifyErr = history.Verify(tx, b, indexesMap)
+		}
+		lock.Unlock()
```
Comment (gz-c) [inline src/visor/db.go]: I think it is better to do `defer lock.Unlock()` for all locks where possible

**Comment by gz-c** [comment]
deadlock failure

File: src/visor/db.go [hunk 4]
```diff
@@ -49,45 +51,160 @@ func CheckDatabase(db *dbutil.DB, pubkey cipher.PubKey, quit chan struct{}) erro
...
 			return err
 		}
 
-		// Verify historydb
-		return history.Verify(tx, b, indexesMap)
+		// Verify historydb, we don't return the error of history.Verify here,
+		// as we have to check all signature, if we return error early here, the
+		// potential bad signature won't be detected.
+		lock.Lock()
+		defer lock.Lock()
```
Comment (gz-c) [inline src/visor/db.go]: lock.Unlock

File: src/visor/db.go [hunk 5]
```diff
@@ -49,45 +51,160 @@ func CheckDatabase(db *dbutil.DB, pubkey cipher.PubKey, quit chan struct{}) erro
...
+
+	type rebuildResult struct {
+		DB  *dbutil.DB
+		Err error
+	}
+
+	resC := make(chan rebuildResult, 1)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
```
Comment (gz-c) [inline src/visor/db.go]: What is this goroutine for? Is something supposed to run parallel with it?

**Pull request merged**